File: source/PortableSettingsProvider/U3SettingsProvider.vb
Public Class U3SettingsProvider
    Inherits PortableSettingsProvider

    Private Const U3APPDATAPATH As String = "U3_APP_DATA_PATH"

    Overrides Function GetAppSettingsPath() As String
        'Get the environment variable set by the U3 application for a pointer to its data
        Try
            Return Environment.GetEnvironmentVariable(U3APPDATAPATH)
        Catch ex As Exception
            'Not running in a U3 environment, just return the application path
            Return MyBase.GetAppSettingsPath
        End Try
    End Function
End Class

File: source/PortableSettingsProvider/U3SettingsProvider.cs
using System;

namespace PortableSettingsProvider
{
    public class U3SettingsProvider : PortableSettingsProvider
    {
        private const string U3APPDATAPATH = "U3_APP_DATA_PATH";

        public override string GetAppSettingsPath()
        {
            // Get the environment variable set by the U3 application for a pointer to its data
            try
            {
                return Environment.GetEnvironmentVariable(U3APPDATAPATH);
            }
            catch (Exception)
            {
                // Not running in a U3 environment, just return the application path

                return base.GetAppSettingsPath();
            }
        }
    }
}
